<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simplex Visualization of Cell Fate Similarity in Single-Cell Data • CytoSimplex</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simplex Visualization of Cell Fate Similarity in Single-Cell Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CytoSimplex</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><a class="dropdown-item" href="../articles/hspc_example.html">HSPC analysis</a></li>
    <li><a class="dropdown-item" href="../articles/basics.html">General usage with BMSC subset</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">API</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/welch-lab/CytoSimplex"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/LabWelch"><span class="fa fab fa-twitter fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simplex Visualization of Cell Fate Similarity in Single-Cell Data</h1>
                        <h4 data-toc-skip class="author">Yichen Wang,
Jialin Liu</h4>
            
            <h4 data-toc-skip class="date">04/23/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/CytoSimplex/blob/HEAD/vignettes/CytoSimplex.Rmd" class="external-link"><code>vignettes/CytoSimplex.Rmd</code></a></small>
      <div class="d-none name"><code>CytoSimplex.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package use simplex barycentric coordinate approach to assist
exploration in the similarity between single cells between selected cell
clusters. We denote a number (2-4) of selected clusters, or groups of
clusters, as vertices. We calculate the similarity between each single
cell and the average point of each vertex. By normalizing the similarity
between each single cell and all specified vertices to a unit sum, we
can derive a barycentric coordinate for each single cell. Visualization
method for binary (2-ended line), ternary (equilateral triangle) and
quaternary (tetrahedron) simplex are developed. The main plotting
functions are <code><a href="../reference/plotBinary.html">plotBinary()</a></code>, <code><a href="../reference/plotTernary.html">plotTernary()</a></code> and
<code><a href="../reference/plotQuaternary.html">plotQuaternary()</a></code>, respectively. Please see full argument
documentation with <code><a href="../reference/plotBinary.html">?plotBinary</a></code>, <code><a href="../reference/plotTernary.html">?plotTernary</a></code>
and <code><a href="../reference/plotQuaternary.html">?plotQuaternary</a></code>. Here, we show some examples for
creating ternary and quaternary plots, which would be useful.</p>
</div>
<div class="section level2">
<h2 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>In this vignette, we use data from
<a href="https://www.nature.com/articles/s41467-023-38034-2" target="_blank" class="external-link">Matsushita
and Liu, <em>Nat. Comm.</em> 2023</a>. The application of this method
was originally used in this publicaiton as well. From the processed and
annotated scRNA-seq data, we took the subset of 50 cells per major cell
type from the raw count matrix and cell type annotation. These are
embedded within this package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/CytoSimplex/" class="external-link">CytoSimplex</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"rnaRaw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Class of `rnaRaw`: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">rnaRaw</span><span class="op">)</span>, <span class="st">", dimension of `rnaRaw`: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rnaRaw</span><span class="op">)</span>, <span class="st">" genes x "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rnaRaw</span><span class="op">)</span>, <span class="st">" cells"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Class of `rnaRaw`: dgCMatrix, dimension of `rnaRaw`: 20243 genes x 250 cells"</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"rnaCluster"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">rnaCluster</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## rnaCluster</span></span>
<span><span class="co">##   CH  ORT   OS   RE Stem </span></span>
<span><span class="co">##   50   50   50   50   50</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="select-top-features-optional-but-recommended">Select top features (optional but recommended)<a class="anchor" aria-label="anchor" href="#select-top-features-optional-but-recommended"></a>
</h2>
<p>Technically, any forms of feature-by-observation matrix is acceptable
for the method we developed, and users are encouraged to explore the
usability of our method with other types of data, even not biological.
However, single-cell transcriptomics data, as provided, usually is of
high dimensionality and contains technical and biological noise. With
testing different approaches of reducing the dimensionality and noise,
we recommend that users select a number of top differentially expressed
genes (DEGs) for each cluster that a vertex represents.</p>
<p>We implemented a fast Wilcoxon rank-sum test method which can be
invoked with function <code><a href="../reference/selectTopFeatures.html">selectTopFeatures()</a></code>. Here, we will
choose the top DEGs for Osteoblast cells (<code>"OS"</code>), Reticular
cells (<code>"RE"</code>) and Chondrocytes (<code>"CH"</code>), as also
shown in the previously mentioned publication. The number of top DEGs
for each cluster is set to 30 (<code>nTop = 30</code>), thus 90 unique
genes are expected to be returned. Alternatively, users can set
<code>returnStats = TRUE</code> to obtain a table of full Wilcoxon
rank-sum test statistics, including the result for all clusters instead
of selected vertices.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OS"</span>, <span class="st">"RE"</span>, <span class="st">"CH"</span><span class="op">)</span></span>
<span><span class="va">geneSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectTopFeatures.html">selectTopFeatures</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, </span>
<span>                                vertices <span class="op">=</span> <span class="va">vertices</span>, nTop <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">geneSelect</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Nrk"     "Eps8l2"  "Mfi2"    "Scin"    "Fam101a" "Sox5"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectTopFeatures.html">selectTopFeatures</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, </span>
<span>                           vertices <span class="op">=</span> <span class="va">vertices</span>, nTop <span class="op">=</span> <span class="fl">30</span>, returnStats <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   feature group    avgExpr     logFC statistic     auc         pval</span></span>
<span><span class="co">## 1     Rp1    CH  0.0000000 0.0000000    5000.0 0.50000          NaN</span></span>
<span><span class="co">## 2   Sox17    CH  0.0000000 0.0000000    5000.0 0.50000          NaN</span></span>
<span><span class="co">## 3  Mrpl15    CH  8.4139185 5.7710321    6948.0 0.69480 7.500435e-08</span></span>
<span><span class="co">## 4  Lypla1    CH  4.6278876 2.5075283    5894.0 0.58940 4.817036e-03</span></span>
<span><span class="co">## 5 Gm37988    CH  0.2568508 0.2568508    5100.0 0.51000 4.659094e-02</span></span>
<span><span class="co">## 6   Tcea1    CH 11.7897501 6.0049962    6559.5 0.65595 2.617836e-04</span></span>
<span><span class="co">##           padj pct_in pct_out</span></span>
<span><span class="co">## 1          NaN      0     0.0</span></span>
<span><span class="co">## 2          NaN      0     0.0</span></span>
<span><span class="co">## 3 6.578978e-07     64    19.0</span></span>
<span><span class="co">## 4 1.081562e-02     36    15.5</span></span>
<span><span class="co">## 5 7.428025e-02      2     0.0</span></span>
<span><span class="co">## 6 8.540867e-04     86    40.5</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="generating-ternary-plots">Generating ternary plots<a class="anchor" aria-label="anchor" href="#generating-ternary-plots"></a>
</h2>
<p><code><a href="../reference/plotTernary.html">plotTernary()</a></code> shows sample similarity in a ternary
simplex – equilateral triangle. The closer a dot, a cell, is to one
vertex, the more similar the cell is to the cell cluster(s) the vertex
represents. We recommend that users select the top marker genes for each
terminal and only use them as the features for calculating the
similarity.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vt.tern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OS"</span>, <span class="st">"RE"</span>, <span class="st">"CH"</span><span class="op">)</span></span>
<span><span class="va">gene.tern</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectTopFeatures.html">selectTopFeatures</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotTernary.html">plotTernary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span>, features <span class="op">=</span> <span class="va">gene.tern</span><span class="op">)</span></span></code></pre></div>
<p><img src="CytoSimplex_files/figure-html/plotTernary-1.png" width="480" style="display: block; margin: auto;"></p>
<blockquote>
<p>The static figure depends on <code>ggplot2</code>, which is widely
used for visualization in R. The binary simplex is plotted normally in
ggplot 2D coordinates, while for the ternary simplex, the barycentric
coordinate is drawn with 2D segments with cartesian coordinate, instead
of implementing a ternary barycentric coordinate system. Users wishing
to add customized alteration should pay attention to this.</p>
</blockquote>
<div class="section level3">
<h3 id="interactive-ternary-plot">Interactive ternary plot<a class="anchor" aria-label="anchor" href="#interactive-ternary-plot"></a>
</h3>
<p>The same figure above can be shown in an interactive panel, powered
by <a href="https://plotly.com/r/" class="external-link">plotly</a>. The interactive plot
allows users to zoom in and out, and hover over the dots to see the cell
names and the values of inferred similarity. To enable this feature,
users only need to add <code>interactive = TRUE</code> in the function
call.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTernary.html">plotTernary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span>, features <span class="op">=</span> <span class="va">gene.tern</span>, </span>
<span>            interactive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>For a rendered instance, please view our documentation website as the
vignette is only allowed to demonstrate minimum content.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="adding-velocity-information-to-ternary-plot">Adding velocity information to ternary plot<a class="anchor" aria-label="anchor" href="#adding-velocity-information-to-ternary-plot"></a>
</h2>
<p>RNA velocity is a quantitative measurement of cellular transitions
from single-cell transcriptomics experiments and reveals transient
cellular dynamics among a heterogeneous cell population
(<a href="https://www.pnas.org/doi/10.1073/pnas.2105859118" target="_blank" class="external-link">Qiao,
PNAS 2021</a>). We implemented a velocity visualization strategy that
could be applied to ternary and quaternary simplex plot. The velocity
information input format must be an N x N graph (sparse matrix, where N
denotes number of cells). We have included a graph that matches with the
cells in the example dataset in this package. This graph is a subset of
the output from
<a href="https://github.com/welch-lab/VeloVAE" target="_black" class="external-link">Python
module <code>veloVAE</code></a>, as part of the processed data from the
publication mentioned at the start.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"rnaVelo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Class of `rnaVelo`: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">rnaVelo</span><span class="op">)</span>, </span>
<span>             <span class="st">", dimension of `rnaVelo`: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rnaVelo</span><span class="op">)</span>, <span class="st">" x "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rnaVelo</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Class of `rnaVelo`: dgCMatrix, dimension of `rnaVelo`: 250 x 250"</span></span></code></pre>
<p>We create a number of square grids in the 2D plain of the ternary
simplex (or cube grids in 3D space of the quaternary simplex), and
aggregate the cells fall into each grid with taking the mean of velocity
towards each of the vertices. Finally, we draw an arrow from the grid
center pointing to each vertex with the length representing the
aggregated mean velocity.</p>
<p>Interactive view with velocity information is also supported. The
aggregated potential value can also be shown when hovering above the
arrows.</p>
<blockquote>
<p>For a rendered instance of interactive visualization, please view our
documentation website as the vignette is only allowed to demonstrate
minimum content.</p>
</blockquote>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTernary.html">plotTernary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span>, </span>
<span>            features <span class="op">=</span> <span class="va">gene.tern</span>, veloGraph <span class="op">=</span> <span class="va">rnaVelo</span><span class="op">)</span></span></code></pre></div>
<p><img src="CytoSimplex_files/figure-html/ternVelo-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="exploration-with-each-cluster">Exploration with each cluster<a class="anchor" aria-label="anchor" href="#exploration-with-each-cluster"></a>
</h2>
<p>An argument <code>splitCluster</code> is supported for all three
plotting functions. By setting <code>splitCluster = TRUE</code>, A list
of plots will be returned, with one containing all cells, and each of
the other sub-plots containing only dots (cells) belonging to one
cluster in the annotation specified.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">ternList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotTernary.html">plotTernary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span>, </span>
<span>                        features <span class="op">=</span> <span class="va">gene.tern</span>, </span>
<span>                        byCluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Stem"</span>, <span class="st">"RE"</span>, <span class="st">"ORT"</span>, <span class="st">"OS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ternList</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Stem" "RE"   "ORT"  "OS"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ternList</span><span class="op">$</span><span class="va">Stem</span> <span class="op">+</span> <span class="va">ternList</span><span class="op">$</span><span class="va">RE</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">ternList</span><span class="op">$</span><span class="va">ORT</span> <span class="op">+</span> <span class="va">ternList</span><span class="op">$</span><span class="va">OS</span><span class="op">)</span></span></code></pre></div>
<p><img src="CytoSimplex_files/figure-html/splitCluster-1.png" width="672" style="display: block; margin: auto;"></p>
<p>As can be seen in the subplots, osteoblast-chondrocyte transitional
(OCT) stem cells (<code>"Stem"</code>) sit closer to osteoblast vertex
while do not tend to be extremely close to any vertex as observed in the
“OS” and “RE” clusters; reticular cells (<code>"RE"</code>) and
osteoblast cells (<code>"OS"</code>) are gathered towards their
corresponding vertices; osteoblast-reticular transitional cells
(<code>"ORT"</code>) distribute across the vertices for the two cell
types. These patterns match with the conclusion in the publication.</p>
<p>Similarly, the velocity layer can also be splitted.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">veloSplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotTernary.html">plotTernary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.tern</span>, </span>
<span>                         features <span class="op">=</span> <span class="va">gene.tern</span>, veloGraph <span class="op">=</span> <span class="va">rnaVelo</span>, </span>
<span>                         byCluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Stem"</span>, <span class="st">"RE"</span>, <span class="st">"ORT"</span>, <span class="st">"OS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">veloSplit</span><span class="op">$</span><span class="va">Stem</span> <span class="op">+</span> <span class="va">veloSplit</span><span class="op">$</span><span class="va">RE</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">veloSplit</span><span class="op">$</span><span class="va">ORT</span> <span class="op">+</span> <span class="va">veloSplit</span><span class="op">$</span><span class="va">OS</span><span class="op">)</span></span></code></pre></div>
<p><img src="CytoSimplex_files/figure-html/ternVeloSplit-1.png" width="672" style="display: block; margin: auto;"></p>
<p>As shown in the subplots, the OCT stem cells has the transitional
potential towards all three terminal cell types; reticular and
osteablast cells are differentiating towards their corresponding cell
types; while the ORT cells have the transition potential towards both
osteoblast and reticular cell types.</p>
</div>
<div class="section level2">
<h2 id="create-quaternary-simplex-plot">Create quaternary simplex plot<a class="anchor" aria-label="anchor" href="#create-quaternary-simplex-plot"></a>
</h2>
<p>For a quaternary simplex, we need one more cluster as a vertex. Here,
we add the cells annotated as osteoblast-reticular transition cells
(<code>"ORT"</code>) into the vertex list. We also add the velocity
information in this example, as it will not be shown by default. Note
that we provide interactive quaternary plot by default.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vt.quat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OS"</span>, <span class="st">"RE"</span>, <span class="st">"CH"</span>, <span class="st">"ORT"</span><span class="op">)</span></span>
<span><span class="va">gene.quat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectTopFeatures.html">selectTopFeatures</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.quat</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotQuaternary.html">plotQuaternary</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.quat</span>, </span>
<span>               features <span class="op">=</span> <span class="va">gene.quat</span>, veloGraph <span class="op">=</span> <span class="va">rnaVelo</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="CytoSimplex_files/figure-html/plotQuaternary-1.png" width="576" style="display: block; margin: auto;"></p>
<blockquote>
<p>The default of <code><a href="../reference/plotQuaternary.html">plotQuaternary()</a></code> returns an interactive
view. For a rendered instance of interactive visualization, please view
our documentation website as the vignette is only allowed to demonstrate
minimum content.</p>
</blockquote>
<p>We have also implemented of GIF image generator that rotates the
tetrahedron rounding the z-axis. Note that package <code>magick</code>
is required for this feature.
(<a href="https://cran.r-project.org/package=magick/vignettes/intro.html#Installing_magick" target="_blank" class="external-link">See
here for how to install <code>magick</code> in detail</a>)</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/writeQuaternaryGIF.html">writeQuaternaryGIF</a></span><span class="op">(</span><span class="va">rnaRaw</span>, clusterVar <span class="op">=</span> <span class="va">rnaCluster</span>, vertices <span class="op">=</span> <span class="va">vt.quat</span>, </span>
<span>                   features <span class="op">=</span> <span class="va">gene.quat</span>, veloGraph <span class="op">=</span> <span class="va">rnaVelo</span>, </span>
<span>                   width <span class="op">=</span> <span class="fl">8</span>, height <span class="op">=</span> <span class="fl">5</span>, res <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/welch-lab/CytoSimplex/refs/heads/main/rotating_tetra.gif" width="480" style="display: block; margin: auto;"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yichen Wang, Jialin Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
